### PR TITLE
use DNS cache for samba and mysql. (fixes  #15469)

### LINKED ISSUE #15469: Kodi doesn't honor <hosts> in advancedsettings.xml
### Describe the bug
Here is a clear and concise description of what the problem is:

`<videodatabase>` and `<musicdatabase>` doesn't care about what is defined in previous `<hosts>` in advancedsettings.xml.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

If a host is defined in advancedsettings it should use it.

## Actual Behavior

It doesn't matter what is defined in `<hosts>`. Kodi doesn't use it anyway.

### To Reproduce

In advancedsettings.xml:
A1. Define a host with the actual/right hostname (in my case "SC-NAS") but wrong IP.
A2. Use the hostname above when defining a mysql database.
A3. Start Kodi.
A4. Kodi should not find the database because of the hostname points to the wrong IP, but it does.
 
B1. Define a host with the wrong hostname (lets say "TEST") but with the right IP.
B2. Use the hostname above when defining a mysql database.
B3. Start Kodi.
B4. Kodi should now find the database because of the hostname points to the right IP, but it doesn't.

### Debuglog

The debuglog can be found here:
Debug with right hostname and wrong IP: https://paste.kodi.tv/erusocarat.kodi
Debug with wrong hostname and right IP: https://paste.kodi.tv/quqokufupu.kodi

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

Also when referring to a `smb://hostname` in sources.xml doesn't care about what is defined under `<hosts>`.

The problem was discovered when Windows 10 removed SMB1 by default and the hostname wasn't autodiscovered in Windows under a period, before it was fixed.

The same problem is in Kodi 17.x

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [x] Windows
 - [ ] Windows UWP

 - Operating system version/name: 10 Education edition (10.0.17763.253)
 - Kodi version: 18.0

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by wsnipex**
## Description

`<hosts> entries` from advancedsettings currently are ignored in samba and mysql

## Motivation and Context

https://github.com/xbmc/xbmc/issues/15469

## How Has This Been Tested?

I only quickly tested samba on linux.
**mysql is untested, since I cnba to setup mysql**, I hope the reporters in the issue will do that

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by arnova** [comment]
@wsnipex: Can you please also add this for Curl-based filesystems (http/ftp/dav/etc.) ? And perhaps upnp as well? That should make it less confusing (and more consistent) from a user's POV.

**Comment by wsnipex** [comment]
I'll look into it, but I'd like confirmation from at least one of the issue reporters that the current state works as expected, specially mysql

**Comment by phunkyfish** [comment]
jenkins build this please

File: xbmc/dbwrappers/mysqldataset.cpp [hunk 1]
```diff
@@ -145,7 +146,13 @@ int MysqlDatabase::connect(bool create_new) {
   if (host.empty() || db.empty())
     return DB_CONNECTION_NONE;
 
-  //CLog::Log(LOGDEBUG, "Connecting to mysql:%s:%s", host.c_str(), db.c_str());
+  std::string resolvedHost;
+  if (CDNSNameCache::Lookup(host, resolvedHost))
+  {
+    CLog::Log(LOGDEBUG, "{} replacing configured host {} with resolved host {}", host,
+              resolvedHost);
```
Comment (howie-f) [inline xbmc/dbwrappers/mysqldataset.cpp:153]: three times `{}` is too much for two logvars. this is exactly where it stuck :)
fix is fine if this is corrected:
`DEBUG <general>: replacing configured host dummy with resolved host 192.168.200.105`

Comment (wsnipex) [reply xbmc/dbwrappers/mysqldataset.cpp]: :) thanks, fixed

**Review by howie-f — APPROVED** [review]
yw

**Comment by arnova** [comment]
> I'll look into it, but I'd like confirmation from at least one of the issue reporters that the current state works as expected, specially mysql

If it turns out to be problematic, at least update the wiki so folks know for what stuff <hosts> applies.

**Comment by wsnipex** [comment]
@arnova I've added curl support, tested with HTTP and HTTPS

**Comment by arnova** [comment]
> @arnova I've added curl support, tested with HTTP and HTTPS

Thanks! Looks good to me. Didn't know Curl could do that natively :-)

File: xbmc/filesystem/CurlFile.h [hunk 2]
```diff
@@ -185,6 +185,7 @@ namespace XFILE
       bool m_allowRetry;
       bool m_verifyPeer = true;
       bool m_failOnError = true;
+      curl_slist* m_dnsCacheList;
```
Comment (ksooo) [inline xbmc/filesystem/CurlFile.h:188]: = nullptr

Then, there is no need to initialize in the ctor(s).

File: xbmc/filesystem/CurlFile.cpp [hunk 3]
```diff
@@ -694,6 +703,33 @@ void CCurlFile::ParseAndCorrectUrl(CURL &url2)
...
+
+    if (entryPort == 0)
+    {
+      if (url2.GetProtocol() == "http" || url2.GetProtocol() == "dav")
+        entryPort = 80;
+      else if (url2.GetProtocol() == "https" || url2.GetProtocol() == "davs")
+        entryPort = 443;
+      else if (url2.GetProtocol() == "ftp")
+        entryPort = 21;
+      else if (url2.GetProtocol() == "ftps")
```
Comment (ksooo) [inline xbmc/filesystem/CurlFile.cpp:721]: Many calls to `url2.GetProtocol()`here. Maybe define an local variable for the value? Even if `url2.GetProtocol()` is cheap and inline and what not today, you never know how implementation will change in the future, performance-wise.

Comment (wsnipex) [reply xbmc/filesystem/CurlFile.cpp]: fixed

File: xbmc/filesystem/CurlFile.h (hunk 2 above)
Comment (wsnipex) [reply xbmc/filesystem/CurlFile.h]: fixed

**Pull request merged**